### PR TITLE
use flash attn fuse cross entropy loss to reduce metric memory usage

### PR DESCRIPTION
**Pull request by cli99**
This PR uses fused cross entropy loss from flash attention in the metric LanguageCrossEntropy (also LanguagePerplexity). 
The current torch.nn.CrossEntropyLoss call needs `6 * seq_len * vocab_size` GPU memory, and can be the bottleneck memory usage when sequence length is long (where act ckpt is probably used). Using  cross entropy loss from flash attn resolves this problem.

Example test model with long sequence and full act ckpt:
with torch loss fn:
image

with flash_attn loss fn
image

### CODE REVIEW

**Comment by dakinggg** [comment]
@cli99 consider doing it as in https://github.com/mosaicml/llm-foundry/pull/575/files, to avoid introducing flash attn as a composer dependency.

File: composer/metrics/nlp.py [hunk 1]
```diff
@@ -97,7 +97,11 @@ def __init__(self, dist_sync_on_step: bool = False, ignore_index: int = -100):
         super().__init__(dist_sync_on_step=dist_sync_on_step)
 
         self.ignore_index = ignore_index
-        self.loss_fn = torch.nn.CrossEntropyLoss(ignore_index=ignore_index, reduction='sum')
+        try:
+            from flash_attn.losses.cross_entropy import CrossEntropyLoss as FusedCrossEntropyLoss
+            self.loss_fn = FusedCrossEntropyLoss(ignore_index=ignore_index, reduction='sum')
+        except ImportError:
+            self.loss_fn = torch.nn.CrossEntropyLoss(ignore_index=ignore_index, reduction='sum')
```
Comment (j316chuck) [inline composer/metrics/nlp.py:104]: Should this live in llm-foundry? CC: @dakinggg

Comment (dakinggg) [reply composer/metrics/nlp.py:104]: see above

Comment (dakinggg) [reply composer/metrics/nlp.py:104]: i have a very old pr that i never merged

**Comment by mvpatel2000** [comment]
@cli99 should we close or is this still WIP
